crates/ra_syntax/src/yellow/syntax_error.rs

```diff
@@ -0,0 +1,42 @@
+use std::fmt;
+
+use crate::TextRange;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SyntaxError {
+    pub kind: SyntaxErrorKind,
+    pub range: TextRange,
```

Comment:
I am slightly worried that we know expose a rather elaborate internal structure of the error. Let's make these fields private, but add

```
enum Location {
    Offest(TextUnit),
    Range(TextRange),
}

impl SyntaxError {
    fn location() -> ErrorLocation
}
```

I think it maybe make sense to distinguish between offset/range for location, because there are errors for which range does not make sense, due to error recovery. Consider, for example, this snippet:

```
fn

fn foo() {}
```

Here, we should report `expected ident`. However, the `fn foo() {}` is correctly parsed as a function, so we can't mark `fn` token as an error. I think the expected behavior is to say that the error happened after the `fn` token, and let it to the UI layer to select the most appropriate presentation, which, in this case, is a single-character range *after* `\n`.

Reply:
None that we don't need `fn message() -> String`, because `SyntaxError: fmt::Display`